packages/zoe/src/contracts/operaConcertTicket.js

```diff
@@ -0,0 +1,120 @@
+import { makeZoeHelpers, defaultAcceptanceMsg } from './helpers/zoeHelpers';
+
+import produceIssuer from '@agoric/ertp';
+
+/*
+  Roles in the arrangement:
+  - Contract creator: describes the contract with:
+    - number of seats, show, date/time of start
+    - expected (ERTP) amount per ticket (we assume all tickets cost the same)
+  - Smart Contract: 
+    - mints the tickets
+    - provides the seats
+  - Auditorium (unique contract seat, usually taken by the contract creator): the person hosting 
+  the Opera show, selling the tickets and getting the payment back
+  - Ticket buyers (contract seat created on demand): 
+    - can see the available opera show seats
+    - can consult the terms
+    - can redeem the zoe invite with the proper payment to get the ticket back
+*/
+
+export const makeContract = harden(async zoe => {
+
+  const zoeService = zoe.getZoeService();
+
+  // Create the internal ticket mint
+  const { issuer, mint, amountMath } = produceIssuer('Opera tickets', 'set');
+  await zoe.addNewIssuer(issuer, 'Auditorium');
+
+  // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
+  const {
+    rejectOffer,
+    checkIfProposal,
+    swap,
+    canTradeWith,
+    getActiveOffers,
+    assertKeywords,
+  } = makeZoeHelpers(zoe);
+  
+  const { 
+    terms: { show, start, count, expectedAmountPerTicket }, 
+    issuerKeywordRecord: {Buyer: buyerIssuer}
+  } = zoe.getInstanceRecord();
+  
+  const availableTicketDescriptionByNumber = new Map(Array(count)
+    .fill()
+    .map((_, i) => [i+1, harden({ show, start, number: i + 1 })]))
+  
+  const salesPayouts = new Set()
+
+  const auditoriumSeat = harden({
+    async getSalesPayment(){
```

Comment:
Instead of having this method for the entity selling the tickets to receive their payout, we should use the Zoe payout mechanism instead. The contract should retain the offerHandle of the original creator, then reallocate all the moola to that offerHandle as it is received. The contract will still have to create new invites to escrow tickets with Zoe so that the contract can reallocate them to the user, but at least we 1) guarantee offer safety for the seller of the tickets, and 2) don't have a duplicate mechanism for payouts. 

I think you are right that this "contract escrows things with Zoe to eventually give them to a user" pattern is going to be reused a lot, and we might want to make it a lot simpler.

Reply:
One idea i had initially was that the Opera might want to get the moolas as they arrive (and not only once at the end like how the payout mechanism requires)
But i realize now that it unnecessarily complicates this first contract, so i'll ignore this requirement for now

Another problem i see with the payout is that the Opera would need to know it advance what it `want`s, but it cannot know in advance how many tickets will be sold and so, it cannot know whether it `want` 0 or 22 or 44 or 66 moolas and invite redeeming will only accept one proposal
If the Opera `want`s 22 moolas and 3 tickets get sold, the 44 leftover moolas can only remain in the contract
If the Opera `want`s 66 moolas and 1 ticket is sold, there is never enough moolas to perform a reallocation of the moolas that satisfied `want:` 66 moolas

i do not know how to solve this problem with Zoe. How to describe a proposal with an unknown extent?
(i'm starting to think that maybe 0 could work with a manual call to `.reallocate` with the final amount, i'll dig into that)

Reply:
An alternative to all this would be:
- the contract mints the tickets + invites and passes all this to the Opera
- the Opera redeems invites with `give ticket want 22 moolas`
- the contract has kept the offerHandles and is able to track which ticket is still available with them
- The Opera uses all the individual payouts

With this alternative, the contract never redeems invites internally and never plays with payouts

Reply:
To expand a bit on this idea. I was initially puzzled by the argument "guarantee offer safety for the seller of the tickets" because i felt that if the contract code was correct, offer safety was guaranteed. But this "if" is literally one of the things being defended against, so that's not an assumption that can be made

I feel like a contract that plays with invites and payouts internally cannot really guarantee offer safety since a bug may lead to invites not being properly redeemed with incorrect proposals/payments or payouts not being recovered and that would lead to assets being stuck in the contract which was one of the things offer safety was supposedly defending against

If i'm stretching this idea, I would go as far as considering that inside a zoe contract it shouldn't be possible to call `zoe.redeem`

Reply:
> Another problem i see with the payout is that the Opera would need to know it advance what it wants, but it cannot know in advance how many tickets will be sold 

Yes, this is a lack of expressiveness that we absolutely need to fix. We started by thinking of this as a "divisible" offer, but then we got into roundoff hell. Instead, we're thinking of this a "multiples": Create the smallest atomic offer: "1 ticket for 1 ticket price" and place it inside a repetition factor of N. It acts like a set of N separate but identical offers, any number of which can be accepted.

Another missing expressiveness we expect to add is exclusive choice: "offer A or offer B but not both".

With one additional element, an "offset", these three together lets us express a piecewise linear approximation to an indifference curve, which is an interesting test of "adequate" generality.

Note of this will happen by Alpha.